### ISSUE TITLE
os-frr / FRR 10.6.1: bgpd rejects 10.5.x peer OPEN (NOTIFICATION+RST) and staticd fails to start

### ISSUE THREAD

**Issue by kk6axf** [opnsense/plugins#5514]
# Runbook: OPNsense Update Breaks BGP Over WireGuard (FRR 10.6.1 Regression)

Sanitized for sharing. Hostnames, IPs, keys, and ASNs replaced with generic placeholders. Keep the unsanitized internal copy for actual operations.

Incident date: 2026-06-20/21. Nodes: HUB (hub VPS), GW (gateway, physical), SPOKE-A (physical), SPOKE-B. Root cause: OPNsense 26.1.x update shipped FRR 10.6.1, which broke BGP while WireGuard stayed up.

Placeholder key:
- `HUB` = hub router (VPS). `GW` = a physical spoke gateway. `SPOKE-A` / `SPOKE-B` = additional spoke routers.
- `TUN_HUB` / `TUN_GW` / `TUN_SPOKE-A` / `TUN_SPOKE-B` = each node's tunnel IP. `<peer-tunnel-ip>` = the tunnel IP of the BGP neighbor.
- `REMOTE_HOST` = a host reachable only across the tunnel. `INTERNAL_ZONE` = an internal DNS zone. `DDNS_NAME` = a dynamic-DNS hostname.
- `<ASN>` = the local BGP AS number. `<WG_PORT>` = the WireGuard listen port. `wgX` = the relevant WireGuard interface.

---

## TL;DR

- WireGuard and firewall were **never** the problem. Tunnels were up the whole time.
- The OPNsense update pushed **FRR 10.6.1**, which (a) on one node came up with bgpd alive but **staticd not started / no frr.log**, and (b) **rejects the BGP OPEN** from FRR 10.5.x peers (NOTIFICATION + RST). Either way BGP drops and every route across the tunnel disappears.
- Fix = get the node back to **FRR 10.5.x** to match the mesh (snapshot / boot environment). Then harden router-ids.
- A separate gotcha: rolling the **hub** back to an old snapshot dropped a post-snapshot WireGuard tunnel (the SPOKE-A instance), breaking that spoke for a different reason (missing config, not version).

## Symptom signature (how to recognize this fast)

- `wg show` healthy: recent handshake, bytes both directions. Tunnel IP pings fine.
- But cross-tunnel destinations (remote subnets, internal DNS) unreachable.
- `vtysh -c "show ip bgp summary"`: peers `Idle`/`Active`, often `MsgRcvd 0`.
- One or more of:
  - `pgrep staticd` returns nothing (should be running if `redistribute static` configured).
  - `/var/log/frr/frr.log` missing.
  - `show bgp neighbor X`: `Opens Rcvd 0`, `Last reset: Notification sent ... OpenSent State`.
  - `show version` differs between the broken node and its peers (10.6.1 vs 10.5.x).

## Diagnostic sequence (paste-safe — tcsh chokes on inline `#` comments and `>`/`()`)

On the affected node:
```
vtysh -c "show version"
vtysh -c "show ip bgp summary"
vtysh -c "show bgp neighbor <peer-ip>"
pgrep -lf 'bgpd|zebra|staticd|mgmtd'
ping -c3 <peer-tunnel-ip>
```
Prove transport vs FRR with a wire capture (FreeBSD has no `any` device — name the wg interface):
```
ifconfig -l | tr ' ' '\n' | grep -i wg
tcpdump -ni wgX -vvv tcp port 179
```
- Peer's OPEN arrives + ACKed, node replies 21-byte NOTIFICATION + RST -> **FRR version reject**, not transport.
- `ping <peer-tunnel-ip>` 100% loss -> **WireGuard transport down** (different root cause — check tunnel config / missing instance).

## What we ruled out (don't waste time re-checking)

- **WireGuard / firewall rules** — handshake completes, TCP/179 connects, so the WG port rules are fine.
- **MTU** — DF ping at 1372 data bytes (1400 on wire) passed both ways. (1450 "Message too long" is a *local* DF rejection, not a path failure — ignore it.)
- **DNS / DDNS** — `DDNS_NAME` resolved correctly; internal-zone failure was a *downstream* effect of the dropped BGP route to the dnsmasq override target.
- **`opnsense-revert -r <release> frr10 os-frr`** — DEAD END at the adjacent release. It also shipped FRR 10.6.1 (only os-frr changed minor version).
- **Plain `service frr restart`** — does NOT re-fork the daemons; watchfrr re-attaches to the same wedged bgpd. Must hard stop + verify new PID, or reboot.

## Fix A — restore FRR to 10.5.x (the real fix)

Pick by node type:

- **Physical / ZFS:** boot environment rollback.
  ```
  bectl list
  bectl activate <pre-update-BE>
  reboot
  vtysh -c "show version"            (expect 10.5.x)
  vtysh -c "show ip bgp summary"
  ```
  Then re-import the node's latest config to recover drift since the BE date (config.xml carries NO package versions, so it will NOT re-pull 10.6.1).

- **VPS:** restore pre-update snapshot (keeps FRR at pre-update version; config-only restore also will not change FRR version).

Confirmed results this incident: GW -> 10.5.3, SPOKE-A -> 10.5.3, HUB -> 10.5.2. All sessions re-established once versions matched.

## Fix B — clear stuck collision after versions match

If a session still won't latch (leftover collision), stagger the re-init from ONE side:
```
vtysh -c "configure terminal" -c "router bgp <ASN>" -c "neighbor <peer> shutdown" -c "end"
   (wait ~15s)
vtysh -c "configure terminal" -c "router bgp <ASN>" -c "no neighbor <peer> shutdown" -c "end"
```

## Hardening — pin BGP router-ids (do on every node)

Both saved configs had blank `<routerid>`, so FRR latched onto dynamic DHCP WAN IPs, which aggravated connection-collision loops during recovery. Pin to stable tunnel IPs:

| Node | Router ID |
|---|---|
| HUB | TUN_HUB |
| GW | TUN_GW |
| SPOKE-B | TUN_SPOKE-B |
| SPOKE-A | TUN_SPOKE-A |

Set in UI: Routing ▸ BGP ▸ General → Router ID → Apply. **Runtime `vtysh` pins do NOT persist** and are wiped by any config apply/import — always set in the UI, and re-set AFTER any config restore.

## Caution: rolling back a node's CONFIG can drop newer tunnels

Rolling the hub VPS to an old snapshot reverted its config too. That snapshot **predated** the SPOKE-A WireGuard tunnel, so that interface was gone (`wg show wgX` -> "Device not configured"), breaking SPOKE-A for a *different* reason (missing config, not FRR version). SPOKE-A's own `ping TUN_HUB` was 100% loss = transport down.

Hub tunnel params to recreate if ever lost (hub side):
- Instance `HUB-wgX-SPOKE-A`, listen port `<WG_PORT>`, MTU 1400, tunnel `TUN_HUB/24` (+ IPv6 /64), disable-routes on.
- Hub shared keypair: pub `<REDACTED-PUBKEY>` / priv `<REDACTED-PRIVKEY>` (store in your secrets manager, NOT here).
- Peer `SPOKE-A`: pub `<REDACTED-PEER-PUBKEY>`, AllowedIPs `0.0.0.0/0, ::/0`.
- WAN firewall pass must cover UDP `<WG_PORT>`.

Lesson: prefer **config restore of the latest known-good config** over a stale snapshot when possible; if you must use a stale snapshot/BE for the OS/FRR, **re-import the latest config.xml on top** to recover drift (won't re-pull the bad FRR).

## When is it safe to run OPNsense updates on these nodes again?

There is no automatic all-clear. Gate it:

1. **Wait for a different FRR build.** This regression is FRR **10.6.1**. Do not update until an OPNsense/os-frr release ships a **different** FRR (10.6.2+, a re-pinned 10.5.x, or an os-frr revision noting an FRR/BGP/staticd fix). Read `System ▸ Firmware ▸ Status` changelog + os-frr plugin changelog before applying.
2. **Track the bug.** Subscribe to the filed issue; "safe" = resolved AND shipped in a version visible in the changelog.
3. **Canary, never fleet-wide.** Update ONE non-critical node (a test box or a portable unit, not the hub) in a maintenance window with failover internet ready. Immediately verify:
   ```
   vtysh -c "show version"
   pgrep -lf staticd            (must be running)
   vtysh -c "show ip bgp summary"   (peers Established, PfxRcd > 0)
   ```
4. **If the canary fails** (staticd missing or won't peer with 10.5.x nodes) -> roll that node back (BE / snapshot) and STOP. Do not proceed.
5. **Only after the canary stays clean** and proves it interoperates with the existing 10.5.x nodes -> roll the rest. BGP needs version compatibility across the mesh; never get stuck mid-migration with incompatible OPENs again.
6. **Always** take a fresh snapshot/boot environment on every node BEFORE any future update. Confirm each node HAS a recent BE/snapshot now.

## Post-incident checklist

- [ ] Hub config restored (brings back the SPOKE-A tunnel + other drift); `show version` still 10.5.2.
- [ ] Re-pin router-ids in UI on all nodes (wiped by restore).
- [ ] `show ip bgp summary` on hub: all spokes Established, PfxRcd > 0.
- [ ] SPOKE-A console: `ping TUN_HUB` ok, BGP Established.
- [ ] From hub: `ping REMOTE_HOST` ok, internal-zone DNS resolves.
- [ ] Bug filed with OPNsense (os-frr plugin); subscribed for fix notification.
- [ ] Hold updates on all BGP nodes until fixed FRR version ships + canary-validated.
